### PR TITLE
Make table output sortable

### PR DESCRIPTION
**Pull request by tharindusathis**
This is an experimental feature. I just wanted to sort the tables by clicking on the headers. 
Current output is like this:

Peek 2022-06-26 14-02

### CODE REVIEW

**Comment by fdncred** [comment]
I think this is interested. I've been wanting to do it for a while and it looks like it's just doing the command again with `| sort-by column-name`, clever. 

The part that I'm not sure that I really like is that it keeps appending `| sort-by column-name` as you click through the columns. Maybe that isn't a big deal? 

Also, I'd like to control-click and be able to do `| sort-by column-name1 column-name2` at some point, but that could be another PR.

**Comment by tharindusathis** [comment]
> The part that I'm not sure that I really like is that it keeps appending `| sort-by column-name` as you click through the columns.

I was thinking to fix it as follows: 
- check if the current command already contain a `| sort-by column-name` at the end
- if so, replace it with a new column name
- if the same column clicked multiple then toggle appending  a `-r` or `--reverse`

@fdncred what do you think?

**Comment by rgwood** [comment]
Hey, thanks for the PR; we've been wanting to implement sorting for a while now.

I think this is going to need a different approach, and some changes are required on the Rust side. Rerunning the command with `sort-by` has 2 problems:
1. The command output might change when you rerun it
2. It's inefficient; rerunning the entire command is going to be slower than just sorting the data

I just implemented https://github.com/nushell/nana/pull/50 to help with this kind of thing. We now store card output in `NanaState.card_cache`; I think the right way to do this is to add a new [command](https://tauri.app/v1/guides/features/command/) that sorts output. Something like:

```rust
#[command]
fn sort_card(card_id: String, sort_column: String, state: State<MyState>) -> Result<String, String> {
// get value out of state.card_cache, sort it, return sorted data serialized to JSON
}
```

Is that something you're interested in/able to implement?

**Comment by tharindusathis** [comment]
@rgwood Yeah. That would be the cleanest solution.

**Comment by rgwood** [comment]
@tharindusathis Are you interested in implementing that approach (including the Rust bits)?

**Comment by tharindusathis** [comment]
> @tharindusathis Are you interested in implementing that approach (including the Rust bits)?

I'm interested, however, it will take some time on my side as I'll be bit busy coming days. So if anyone else able that would be great too.

**Comment by rgwood** [comment]
Cool. The sorting should probably use the existing sorting code: https://github.com/nushell/nushell/blob/4d854f36af03dcfe82931ef82dfe274842dc39bc/crates/nu-command/src/filters/sort.rs#L186

I might have time to work on this today but can't promise anything either.

**Comment by tharindusathis** [comment]
> Some work was needed to expose sorting functionality to Nana: [nushell/nushell#5893](https://github.com/nushell/nushell/pull/5893)

Great! I will continue working on implementing the rest.

**Comment by tharindusathis** [comment]
@rgwood I just re-implement the sorting by creating the `sort_card` command from Rust side. Could you please review (specially the Rust code as I'm bit new to Rust). 

New look:
Peek 2022-06-30 22-31

**Comment by fdncred** [comment]
very nice!

**Comment by rgwood** [comment]
The Rust side of this looks good. Please run `cargo fmt --all` in the `src-tauri` directory to format the code.

File: src/app/Output.tsx [hunk 1]
```diff
@@ -99,7 +132,13 @@ const List = ({ value: { vals } }: { value: any }): any => {
   return [];
 };
 
-export const Output = ({ value }: { value: any }): any => {
+export const Output = ({
+  value,
+  onSortOuput,
```
Comment (rgwood) [inline src/app/Output.tsx:137]:
```suggestion
  onSortOutput,
```

**Comment by rgwood** [comment]
Thank you for implementing this! Just a few suggestions (`cargo fmt` and a typo).

I think we can add the sort direction icon in another PR.

**Comment by tharindusathis** [comment]
Thanks @rgwood. Changes after review are added.

**Comment by fdncred** [comment]
nice work. let's land this.

**Comment by fdncred** [comment]
Error building :(
```
   Compiling nana v0.1.0 (C:\Users\username\source\repos\forks\nana\src-tauri)
error[E0432]: unresolved import `nu_command::sort_value`
 --> src\main.rs:8:5
  |
8 | use nu_command::sort_value;
  |     ^^^^^^^^^^^^^^^^^^^^^^ no `sort_value` in the root
```
I'll create a PR.

**Comment by tharindusathis** [comment]
> Error building :(

You need to run `cargo update` in src-tauri  directory before build.

**Comment by fdncred** [comment]
> You need to run cargo update in src-tauri directory before buil

right, that's what i did when i submitted the PR to fix this an hour or so ago. 😆

**Comment by tharindusathis** [comment]
> > You need to run cargo update in src-tauri directory before buil
> 
> right, that's what i did when i submitted the PR to fix this an hour or so ago. laughing

Yeahh. Thanks for the update

**Pull request merged**